In [113]:
import warnings
warnings.filterwarnings('ignore',category=FutureWarning)
import pandas as pd
import os
import seaborn as sns
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score
import numpy as np
import matplotlib.pyplot as plt

In [114]:
d1 = pd.read_excel(r"C:\Users\DELL\Desktop\course_prac\ml\Regression_Projects\Refinance_Project\Weekly_Refinance_Volumes_Data.xlsx")

In [115]:
d1.head()

,Date,Week No,Month No,Year,Refinance,Mortgage Rate,Inflation,Housing Price Index,Treasury Yield,Unemployment Rate,GDP,Business Confidence Index,Consumer Confidence Index,Initial Unemployment Claim,Disposable Income
0,2018-12-31,1,1,2019,885.00,4.51,1.60,187.90,1.70,4.00,2.40,101.01,101.00,329495.25,16027.80
1,2019-01-07,2,1,2019,1094.00,4.51,1.60,187.90,2.16,4.00,2.40,101.01,101.00,348680.14,16027.80
2,2019-01-14,3,1,2019,1078.00,4.51,1.60,187.90,2.20,4.00,2.40,101.01,101.00,322446.86,16027.80
3,2019-01-21,4,1,2019,1032.00,4.51,1.60,187.90,1.75,4.00,2.40,101.01,101.00,264000.71,16027.80
4,2019-01-28,5,2,2019,989.00,4.51,1.50,188.84,2.17,3.80,2.40,101.07,101.06,251632.29,16073.80


In [116]:
d1.tail()

,Date,Week No,Month No,Year,Refinance,Mortgage Rate,Inflation,Housing Price Index,Treasury Yield,Unemployment Rate,GDP,Business Confidence Index,Consumer Confidence Index,Initial Unemployment Claim,Disposable Income
196,2022-10-02,40,10,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06
197,2022-10-09,41,10,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06
198,2022-10-16,42,10,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06
199,2022-10-23,43,10,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06
200,2022-10-30,44,10,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06


In [117]:
d1.shape

(201, 15)

In [118]:
d1.columns

Index(['Date', 'Week No', 'Month No', 'Year', 'Refinance', 'Mortgage Rate',
       'Inflation', 'Housing Price Index', 'Treasury Yield',
       'Unemployment Rate', 'GDP', 'Business Confidence Index',
       'Consumer Confidence Index', 'Initial Unemployment Claim',
       'Disposable Income'],
      dtype='object')

In [119]:
d1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 15 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Date                        201 non-null    datetime64[ns]
 1   Week No                     201 non-null    int64         
 2   Month No                    201 non-null    int64         
 3   Year                        201 non-null    int64         
 4   Refinance                   188 non-null    float64       
 5   Mortgage Rate               201 non-null    float64       
 6   Inflation                   201 non-null    float64       
 7   Housing Price Index         201 non-null    float64       
 8   Treasury Yield              201 non-null    float64       
 9   Unemployment Rate           201 non-null    float64       
 10  GDP                         201 non-null    float64       
 11  Business Confidence Index   201 non-null    float64       

In [120]:
d1.describe().T

,count,mean,min,25%,50%,75%,max,std
Date,201,2020-11-29 23:24:10.746268672,2018-12-31 00:00:00,2019-12-16 00:00:00,2020-11-30 00:00:00,2021-11-15 00:00:00,2022-10-30 00:00:00,NaN
Week No,201.00,25.76,1.00,13.00,26.00,38.00,53.00,14.75
Month No,201.00,6.35,1.00,3.00,6.00,9.00,12.00,3.37
Year,201.00,2020.44,2019.00,2019.00,2020.00,2021.00,2022.00,1.10
Refinance,188.00,2424.27,464.00,1729.50,2514.00,2953.50,4636.00,963.11
Mortgage Rate,201.00,3.66,2.65,2.98,3.49,4.08,5.81,0.85
Inflation,201.00,3.93,0.10,1.50,2.30,6.80,9.10,3.03
Housing Price Index,201.00,226.88,187.90,195.66,217.84,254.16,282.69,31.32
Treasury Yield,201.00,1.61,0.76,1.17,1.59,1.94,3.01,0.51
Unemployment Rate,201.00,5.27,3.50,3.60,3.90,6.00,14.70,2.61


In [121]:
d1.corr()

,Date,Week No,Month No,Year,Refinance,Mortgage Rate,Inflation,Housing Price Index,Treasury Yield,Unemployment Rate,GDP,Business Confidence Index,Consumer Confidence Index,Initial Unemployment Claim,Disposable Income
Date,1.00,0.17,0.17,0.97,-0.12,0.23,0.87,0.97,0.32,-0.10,0.08,-0.92,0.35,-0.13,0.74
Week No,0.17,1.00,1.00,-0.08,0.11,-0.09,0.07,0.10,-0.20,-0.01,0.24,-0.14,-0.29,-0.11,-0.05
Month No,0.17,1.00,1.00,-0.08,0.11,-0.10,0.08,0.11,-0.20,-0.02,0.25,-0.14,-0.30,-0.12,-0.04
Year,0.97,-0.08,-0.08,1.00,-0.15,0.25,0.85,0.95,0.37,-0.10,0.02,-0.89,0.43,-0.10,0.76
Refinance,-0.12,0.11,0.11,-0.15,1.00,-0.79,-0.54,-0.32,-0.80,0.64,-0.08,0.11,-0.54,0.52,0.16
Mortgage Rate,0.23,-0.09,-0.10,0.25,-0.79,1.00,0.54,0.34,0.80,-0.45,-0.11,-0.29,0.60,-0.26,-0.06
Inflation,0.87,0.07,0.08,0.85,-0.54,0.54,1.00,0.94,0.63,-0.47,0.06,-0.76,0.57,-0.40,0.48
Housing Price Index,0.97,0.10,0.11,0.95,-0.32,0.34,0.94,1.00,0.47,-0.24,0.08,-0.89,0.46,-0.24,0.67
Treasury Yield,0.32,-0.20,-0.20,0.37,-0.80,0.80,0.63,0.47,1.00,-0.61,-0.00,-0.28,0.73,-0.48,0.04
Unemployment Rate,-0.10,-0.01,-0.02,-0.10,0.64,-0.45,-0.47,-0.24,-0.61,1.00,-0.26,-0.10,-0.45,0.84,0.29


In [122]:
future_data = d1[d1['Refinance'].isna()]

In [123]:
future_data.head()

,Date,Week No,Month No,Year,Refinance,Mortgage Rate,Inflation,Housing Price Index,Treasury Yield,Unemployment Rate,GDP,Business Confidence Index,Consumer Confidence Index,Initial Unemployment Claim,Disposable Income
188,2022-08-08,32,8,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06
189,2022-08-15,33,8,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06
190,2022-08-22,34,8,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06
191,2022-08-29,35,8,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06
192,2022-09-05,36,9,2022,NaN,5.41,9.10,274.25,2.38,3.60,2.50,96.60,105.14,213138.56,18423.06


In [124]:
future_data.shape

(13, 15)

In [125]:
base_data = d1[~d1['Refinance'].isna()]

In [126]:
base_data.shape

(188, 15)

In [127]:
base_data.isna().sum()

Date                          0
Week No                       0
Month No                      0
Year                          0
Refinance                     0
Mortgage Rate                 0
Inflation                     0
Housing Price Index           0
Treasury Yield                0
Unemployment Rate             0
GDP                           0
Business Confidence Index     0
Consumer Confidence Index     0
Initial Unemployment Claim    0
Disposable Income             0
dtype: int64

In [128]:
base_data = base_data.drop('Date',axis = 1)

In [129]:
base_data.head()

,Week No,Month No,Year,Refinance,Mortgage Rate,Inflation,Housing Price Index,Treasury Yield,Unemployment Rate,GDP,Business Confidence Index,Consumer Confidence Index,Initial Unemployment Claim,Disposable Income
0,1,1,2019,885.00,4.51,1.60,187.90,1.70,4.00,2.40,101.01,101.00,329495.25,16027.80
1,2,1,2019,1094.00,4.51,1.60,187.90,2.16,4.00,2.40,101.01,101.00,348680.14,16027.80
2,3,1,2019,1078.00,4.51,1.60,187.90,2.20,4.00,2.40,101.01,101.00,322446.86,16027.80
3,4,1,2019,1032.00,4.51,1.60,187.90,1.75,4.00,2.40,101.01,101.00,264000.71,16027.80
4,5,2,2019,989.00,4.51,1.50,188.84,2.17,3.80,2.40,101.07,101.06,251632.29,16073.80


In [130]:
x = base_data.drop('Refinance',axis = 1)
y = base_data['Refinance']

In [131]:
x.head()

,Week No,Month No,Year,Mortgage Rate,Inflation,Housing Price Index,Treasury Yield,Unemployment Rate,GDP,Business Confidence Index,Consumer Confidence Index,Initial Unemployment Claim,Disposable Income
0,1,1,2019,4.51,1.60,187.90,1.70,4.00,2.40,101.01,101.00,329495.25,16027.80
1,2,1,2019,4.51,1.60,187.90,2.16,4.00,2.40,101.01,101.00,348680.14,16027.80
2,3,1,2019,4.51,1.60,187.90,2.20,4.00,2.40,101.01,101.00,322446.86,16027.80
3,4,1,2019,4.51,1.60,187.90,1.75,4.00,2.40,101.01,101.00,264000.71,16027.80
4,5,2,2019,4.51,1.50,188.84,2.17,3.80,2.40,101.07,101.06,251632.29,16073.80


In [132]:
y.head()

0    885.00
1   1094.00
2   1078.00
3   1032.00
4    989.00
Name: Refinance, dtype: float64

In [133]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(x,y,test_size=0.05,random_state=42)

In [134]:
print(X_train.shape,X_test.shape,y_train.shape,y_test.shape)

(178, 13) (10, 13) (178,) (10,)


In [135]:
X_train

,Week No,Month No,Year,Mortgage Rate,Inflation,Housing Price Index,Treasury Yield,Unemployment Rate,GDP,Business Confidence Index,Consumer Confidence Index,Initial Unemployment Claim,Disposable Income
60,9,2,2020,3.49,2.30,198.31,1.27,3.50,-5.10,101.18,101.18,203910.00,16734.80
123,19,5,2021,2.98,5.00,239.04,1.69,5.80,6.70,99.83,99.76,481335.86,17980.60
42,43,10,2019,3.69,1.80,195.30,1.62,3.60,1.90,101.11,101.11,190172.29,16373.60
155,51,12,2021,3.10,7.00,256.67,1.07,3.90,6.90,97.73,97.78,255835.00,18329.50
55,4,1,2020,3.65,2.50,196.57,1.25,3.50,-5.10,101.63,101.64,266356.29,16622.60
...,...,...,...,...,...,...,...,...,...,...,...,...,...
106,2,1,2021,2.67,1.40,219.90,1.33,6.40,6.30,99.10,98.99,1041158.00,19120.30
14,15,4,2019,4.08,2.00,191.98,2.09,3.60,3.20,101.46,101.46,196196.57,16110.10
92,41,10,2020,2.88,1.20,214.57,1.13,6.90,4.50,99.14,99.10,752045.00,17398.90
179,23,6,2022,5.23,9.10,272.21,2.20,3.60,2.50,96.70,98.70,206099.00,18388.74


In [136]:
y

0      885.00
1     1094.00
2     1078.00
3     1032.00
4      989.00
        ...  
183    464.00
184    628.00
185    794.00
186    805.00
187    891.00
Name: Refinance, Length: 188, dtype: float64

In [137]:
y_test

183    464.00
162   2068.00
18    1691.00
15    1789.00
67    4325.00
108   4636.00
45    2325.00
76    3411.00
16    1967.00
146   2805.00
Name: Refinance, dtype: float64

In [138]:
from sklearn.linear_model import LinearRegression
ln=LinearRegression()
ln.fit(X_train,y_train)   #-- Training the model with training data
ln.predict(X_test)

array([1126.74075232, 2097.58964429, 1671.63418666, 1761.48895014,
       3860.3637722 , 3294.45880973, 2291.16983041, 3489.47023178,
       1471.02406654, 2425.32370133])

In [139]:
abs(y_test.values-ln.predict(X_test)).sum() 

np.float64(3533.337311377516)

In [140]:
from sklearn.linear_model import Lasso    #---  L1 Regularisation --
ls=Lasso()
ls.fit(X_train,y_train)
ls.predict(X_test)


C:\Users\DELL\AppData\Roaming\Python\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.436e+07, tolerance: 1.587e+04
  model = cd_fast.enet_coordinate_descent(


array([1109.70691656, 2091.63701083, 1676.68679734, 1770.36556047,
       3868.4257088 , 3297.02091981, 2286.53003221, 3484.87867355,
       1486.47972372, 2432.13234876])

In [141]:
abs(y_test.values-ls.predict(X_test)).sum()

np.float64(3463.5815098297317)

In [142]:
from sklearn.linear_model import Ridge # -- L2 Regularisation
rd=Ridge()
rd.fit(X_train,y_train)
rd.predict(X_test)



array([1093.72967937, 2084.54530092, 1687.03833616, 1781.87989434,
       3874.88600481, 3298.47370123, 2277.13792231, 3487.51015187,
       1501.86714445, 2436.38937579])

In [143]:
abs(y_test.values-rd.predict(X_test)).sum()

np.float64(3403.112753076246)

In [144]:
def table_validation():
  validation_table = []
  for i in [LinearRegression,Lasso,Ridge]:
    rd = i()
    rd.fit(X_train,y_train)
    rd.predict(X_test)

    mae = mean_absolute_error(y_test.values,rd.predict(X_test))
    mse = mean_squared_error(y_test.values,rd.predict(X_test))
    rmse = np.sqrt(mean_squared_error(y_test.values,rd.predict(X_test)))
    mape=(((abs(y_test.values-rd.predict(X_test))/y_test.values)*100).sum())*(1/y_test.shape[0])
    rsquare=r2_score(y_test.values,rd.predict(X_test))
    validation_table.append([i,mae,mse,rmse,mape,rsquare])

  validation_table = pd.DataFrame(validation_table)
  validation_table.columns = ['model name','mae','mse','rmse','mape','rsquare']
  return validation_table

In [145]:
tv = table_validation()

C:\Users\DELL\AppData\Roaming\Python\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.436e+07, tolerance: 1.587e+04
  model = cd_fast.enet_coordinate_descent(


In [146]:
tv

,model name,mae,mse,rmse,mape,rsquare
0,<class 'sklearn.linear_model._base.LinearRegre...,353.33,285430.06,534.26,22.91,0.80
1,<class 'sklearn.linear_model._coordinate_desce...,346.36,279624.13,528.79,22.32,0.81
2,<class 'sklearn.linear_model._ridge.Ridge'>,340.31,274884.58,524.29,21.75,0.81


In [147]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import BaggingRegressor

In [148]:
def model_validation(method,X_train,X_test,y_train,y_test):
    validation_table=[]
    models=[LinearRegression,Lasso,Ridge,DecisionTreeRegressor,RandomForestRegressor,BaggingRegressor]
    for i in models:
        rd=i()
        rd.fit(X_train,y_train)
        if method=="train":
            rd.predict(X_train)
            mae=mean_absolute_error(y_train.values,rd.predict(X_train))
            mse=mean_squared_error(y_train.values,rd.predict(X_train))
            rmse=np.sqrt(mean_squared_error(y_train.values,rd.predict(X_train)))
            mape=(((abs(y_train.values-rd.predict(X_train))/y_train.values)*100).sum())*(1/y_train.shape[0])
            rsquare=r2_score(y_train.values,rd.predict(X_train))
            validation_table.append([i,mae,mse,rmse,mape,rsquare])
        
        
        else:  #-- Predict
            rd.predict(X_test)
            mae=mean_absolute_error(y_test.values,rd.predict(X_test))
            mse=mean_squared_error(y_test.values,rd.predict(X_test))
            rmse=np.sqrt(mean_squared_error(y_test.values,rd.predict(X_test)))
            mape=(((abs(y_test.values-rd.predict(X_test))/y_test.values)*100).sum())*(1/y_test.shape[0])
            rsquare=r2_score(y_test.values,rd.predict(X_test))
            validation_table.append([i,mae,mse,rmse,mape,rsquare])
            
    validation_table=pd.DataFrame(validation_table)
    validation_table.columns=["model name","mae","mse","rmse","mape","rsquare"]
    return validation_table

In [149]:
pd.options.display.float_format = '{:0.2f}'.format

In [150]:
validation_table_train = model_validation('train',X_train,X_test,y_train,y_test)

C:\Users\DELL\AppData\Roaming\Python\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.436e+07, tolerance: 1.587e+04
  model = cd_fast.enet_coordinate_descent(


In [151]:
validation_table_train

,model name,mae,mse,rmse,mape,rsquare
0,<class 'sklearn.linear_model._base.LinearRegre...,319.12,173506.21,416.54,17.03,0.81
1,<class 'sklearn.linear_model._coordinate_desce...,319.23,173567.94,416.61,17.05,0.81
2,<class 'sklearn.linear_model._ridge.Ridge'>,320.51,173833.87,416.93,17.14,0.81
3,<class 'sklearn.tree._classes.DecisionTreeRegr...,0.00,0.00,0.00,0.00,1.00
4,<class 'sklearn.ensemble._forest.RandomForestR...,91.11,18839.45,137.26,4.58,0.98
5,<class 'sklearn.ensemble._bagging.BaggingRegre...,98.81,27958.15,167.21,5.13,0.97


In [152]:
validation_table_test = model_validation('test',X_train,X_test,y_train,y_test)

C:\Users\DELL\AppData\Roaming\Python\Python313\site-packages\sklearn\linear_model\_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.436e+07, tolerance: 1.587e+04
  model = cd_fast.enet_coordinate_descent(


In [153]:
validation_table_test

,model name,mae,mse,rmse,mape,rsquare
0,<class 'sklearn.linear_model._base.LinearRegre...,353.33,285430.06,534.26,22.91,0.80
1,<class 'sklearn.linear_model._coordinate_desce...,346.36,279624.13,528.79,22.32,0.81
2,<class 'sklearn.linear_model._ridge.Ridge'>,340.31,274884.58,524.29,21.75,0.81
3,<class 'sklearn.tree._classes.DecisionTreeRegr...,253.10,154591.90,393.18,8.96,0.89
4,<class 'sklearn.ensemble._forest.RandomForestR...,245.17,129129.45,359.35,11.71,0.91
5,<class 'sklearn.ensemble._bagging.BaggingRegre...,256.07,118232.25,343.85,12.18,0.92


### Model Evaluation Summary

#### Linear Regression, Lasso & Ridge
- Train and Test R² are around **0.80 - 0.81**.
- Performance is consistent on both datasets, indicating **good generalization**.
- However, the overall predictive power is moderate, suggesting that the relationship may be non-linear.
- Hyperparameter tuning will be performed to check for further improvements.

#### Decision Tree Regressor
- Train R² = **1.00** and Train MSE = **0**, indicating the model has perfectly memorized the training data.
- Test R² dropped to **0.89**, showing **significant overfitting**.
- Hyperparameter tuning is required to control model complexity and improve generalization.

#### Random Forest Regressor
- Train R² = **0.98** and Test R² = **0.91**.
- Training error is much lower than testing error, indicating **mild overfitting**.
- Hyperparameter tuning is required to optimize model performance and reduce overfitting.

#### Bagging Regressor
- Train R² = **0.97** and Test R² = **0.92**.
- The model generalizes better than Decision Tree and Random Forest but still shows a small train-test gap.
- Hyperparameter tuning will be performed to further improve robustness and predictive accuracy.

### Conclusion
Tree-based models outperform linear models on this dataset. However, the presence of train-test performance gaps indicates the need for **Hyperparameter Tuning** to obtain the optimal model configuration and improve generalization on unseen data.

## The models are working good on the training data but not working well on the testing data, so we need to do further Processing
##### we need to check multicollinearity, if the features are highly correlated, we need to remove one feature from the highly correlated features
##### By Using Lasso regression model , we can check the features coefficients values, if the coeffis having 0 values, we can remove them from the model building
